plugins/outputs/otlp/client.go

```diff
@@ -0,0 +1,314 @@
+)
+
+// Client allows reading and writing from/to a remote gRPC endpoint. The
+// implementation may hit a single backend, so the application should create a
+// number of these clients.
+type Client struct {
+	logger           log.Logger
+	url              *url.URL
+	timeout          time.Duration
+	rootCertificates []string
+	headers          metadata.MD
+	compressor       string
+
+	conn *grpc.ClientConn
+}
+
+// ClientConfig configures a Client.
+type ClientConfig struct {
+	Logger           log.Logger
+	URL              *url.URL
+	Timeout          time.Duration
+	RootCertificates []string
+	Headers          metadata.MD
+	Compressor       string
+}
+
+// NewClient creates a new Client.
+func NewClient(conf ClientConfig) *Client {
+	logger := conf.Logger
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &Client{
+		logger:           logger,
+		url:              conf.URL,
+		timeout:          conf.Timeout,
+		rootCertificates: conf.RootCertificates,
+		headers:          conf.Headers,
+		compressor:       conf.Compressor,
+	}
+}
+
+// getConnection will dial a new connection if one is not set.  When
+// dialing, this function uses its a new context and the same timeout
+// used for Store().
+func (c *Client) getConnection(ctx context.Context) (_ *grpc.ClientConn, retErr error) {
+	if c.conn != nil {
+		return c.conn, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	useAuth := c.url.Scheme != "http"
+	_ = level.Debug(c.logger).Log(
+		"msg", "new OTLP connection",
+		"auth", useAuth,
+		"url", c.url.String(),
+		"timeout", c.timeout)
+
+	dopts := []grpc.DialOption{
+		grpc.WithBlock(), // Wait for the connection to be established before using it.
+		grpc.WithDefaultServiceConfig(serviceConfig),
+	}
+	if useAuth {
+		var tcfg tls.Config
+		if len(c.rootCertificates) != 0 {
+			certPool := x509.NewCertPool()
+
+			for _, cert := range c.rootCertificates {
+				bs, err := ioutil.ReadFile(cert)
+				if err != nil {
+					return nil, fmt.Errorf("could not read certificate authority certificate: %s: %w", cert, err)
+				}
+
+				ok := certPool.AppendCertsFromPEM(bs)
+				if !ok {
+					return nil, fmt.Errorf("could not parse certificate authority certificate: %s: %w", cert, err)
+				}
+			}
+
+			tcfg = tls.Config{
+				ServerName: c.url.Hostname(),
+				RootCAs:    certPool,
+			}
+		}
+		_ = level.Debug(c.logger).Log(
+			"msg", "TLS configured",
+			"server", c.url.Hostname(),
+			"root_certs", fmt.Sprint(c.rootCertificates),
+		)
+		dopts = append(dopts, grpc.WithTransportCredentials(credentials.NewTLS(&tcfg)))
+	} else {
+		dopts = append(dopts, grpc.WithInsecure())
+	}
+	if c.compressor != "" && c.compressor != "none" {
+		dopts = append(dopts, grpc.WithDefaultCallOptions(grpc.UseCompressor(c.compressor)))
+	}
+	address := c.url.Hostname()
+	if len(c.url.Port()) > 0 {
+		address = net.JoinHostPort(address, c.url.Port())
+	}
+	conn, err := grpc.DialContext(ctx, address, dopts...)
+	c.conn = conn
```

Comment:
Please check the error first before assigning the connection.

Reply:
Done.